cfpb/sbl-frontend
# feat(env): load env vars at runtime with import-meta-env

Allows env variables to be injected in `sbl-frontend` at runtime by using `import-meta-env` and a few other tweaks to nginx, which allows for a bunch of fun stuff:
- configmaps can be used to set env vars (https://github.com/cfpb/regtech-deployments/pull/15)
- you can bounce a container to refresh env vars after modifying its configmap
- you can run `yarn docker-build-and-run` to run just the frontend with env vars injected
- no changes to local dev env work (keep using `.env` file like normal)
- add `sbl-frontend` to `sbl-project` `docker-compose.yaml` (https://github.com/cfpb/sbl-project/pull/211)
- alert users when they're missing env vars

There's a lot more that can be done here (see Future Improvements section), but this should be a good start. 🤞

This PR changes quite a few things, so I listed why I made each change in the section below (starting from the top of files changed list and working my way down) and included a bunch of code comments which I hope is helpful. 🤞

## Changes
- add `.env` to `.dockerignore` so that `.env` changes don't trigger image rebuilds
- add `.env.example` for use by `import-meta-env` to define required env variables and to make it easy for new users to copy it and create their own `.env` files
- modify `.github/workflows/test.yml` to add env variables to github workflows
- modify `.gitignore` to ignore all .env files except the new `.env.example`
- install new packages
  - `import-meta-env` for injecting env vars, flexibility between `.env` and env variables being passed on the command line, allowing configmaps
    - includes: `@import-meta-env/cli`, `@import-meta-env/unplugin`, `dotenv`
  - `dotenv-cli` for automating env var injection for `yarn docker-run` command
 - modify Dockerfile for env var injection
   - add `import-meta-env`
   - copy nginx config to allow for env var injection
   - change `CMD` to `ENTRYPOINT` for env var injection and switching to non-root user
 - update env-guide with instructions on adding new env vars
 - update index.html with a placeholder for injecting env vars for prod builds / cache busting
 - add nginx-entrypoint.sh for env var injection and switching to non-root user
 - allow for env vars in `nginx.conf` to modify the CSP to add `localhost` when serving from docker locally 
 - modify package.json
   - add new packages (see above)
   - add a few `sbl-frontend` only ways of running docker (in the future, we'll more often than not use `sbl-project`, see: https://github.com/cfpb/sbl-project/pull/211)
   - modify `docker-run` to work now
     - removed the `&& open http://localhost:8080` to `docker-run` because it caused weird race conditions and made it hard to kill the running container
     - changed the port to `8085` to avoid using the same port as other containers
   - add `docker-build-and-run` to combine build and run commands
 - modify `axiosService.ts` to change from a proxy based approach for base urls to per API axios instance approach. I talked with @shindigira about this: trying to replicate this kind of behavior in nginx causes some load balancer / upload issues, and causes some weirdness. This is a commonly used approach around these issues.
 - modify `common.ts` to include `MAIL_BASE_URL`
 - lint `oidc.ts` suggested removing `as string` now, since these env vars are [already typed string](https://github.com/cfpb/sbl-frontend/blob/main/src/vite-env.d.ts)
 - all API calls in `/requests` now use an axios instance for their API calls instead of relying on the proxy
 - modify `vite.config.ts`
   - remove proxy
   - add import-meta-env
   - remove `envPrefix` since import-meta-env handles that now

## Future improvements
- https://github.com/cfpb/sbl-frontend/issues/589
- https://github.com/cfpb/sbl-frontend/issues/594
- https://github.com/cfpb/sbl-frontend/issues/595
- https://github.com/cfpb/sbl-frontend/issues/426

## How to test

1. If you're missing a variable in your `.env` file (see [.env.example](https://github.com/cfpb/sbl-frontend/pull/428/files#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8c) for complete list) do you see an error appear?

![Screenshot 2024-05-29 at 7 00 36 AM](https://github.com/cfpb/sbl-frontend/assets/19983248/28933085-17b6-4f58-adcd-4bfc6b0ef55e)


2. Are the github actions tests on this PR that now run on env vars passing?

<img width="932" alt="Screenshot 2024-05-29 at 6 56 49 AM" src="https://github.com/cfpb/sbl-frontend/assets/19983248/af48b49f-4b89-4533-be6a-77a93b2b6c9f">


3. Can you run `yarn docker-build` and `yarn docker-run` and see the website appear on http://localhost:8085/?
  - You may need to refresh the page after everything spins up, also see a future PR: [#589](https://github.com/cfpb/sbl-frontend/issues/589)
  - We'll mostly be using `sbl-project` [to spin up](https://github.com/cfpb/sbl-project/pull/211) the `sbl-frontend` container, but this is nice to have for standalone testing
<img width="830" alt="Screenshot 2024-05-29 at 7 31 42 AM" src="https://github.com/cfpb/sbl-frontend/assets/19983248/902923de-5d0b-480b-a051-0abd867b5923">


## Related PRs
- https://github.com/cfpb/sbl-project/pull/211
- https://github.com/cfpb/regtech-deployments/pull/15
- enterprise pr
